src/ai/defaultai_seafaring.cc

```diff
@@ -227,21 +225,38 @@ bool DefaultAI::marine_main_decisions(const Time& gametime) {
 		}
 	}
 
+	if (game().naval_warfare_allowed() && ports_finished_count * 2 > warships_count &&
+	    tradeships_count > expeditions_in_progress + ports_count / 3 + 1) {
+		warship_needed = true;
+	}
+	if (!ship_free && warships_count > 0) {
+		tradeship_refit_needed = true;
+	}
+
 	// starting an expedition? if yes, find a port and order it to start an expedition
+	bool start_expedition = false;
 	if (ports_count > 0 && expeditions_in_progress == 0 && expeditions_in_prep == 0 &&
 	    !persistent_data->no_more_expeditions && ship_free && basic_economy_established) {
+		start_expedition = true;
+	}
 
-		// we need to find a port
-		for (const WarehouseSiteObserver& wh_obs : warehousesites) {
-			if (wh_obs.bo->is(BuildingAttribute::kPort)) {
-				verb_log_dbg_time(game().get_gametime(),
-				                  "  %1d: Starting preparation for expedition in port at %3dx%3d\n",
-				                  player_number(), wh_obs.site->get_position().x,
-				                  wh_obs.site->get_position().y);
-				game().send_player_start_or_cancel_expedition(*wh_obs.site);
-				return true;
-			}
+	// handle ports: start expedition and decide whether we need guard ships.
+	for (PortSiteObserver& p_obs : portsites) {
+		if (start_expedition) {
+			verb_log_dbg_time(game().get_gametime(),
+			                  "  %1d: Starting preparation for expedition in port at %3dx%3d\n",
+			                  player_number(), p_obs.site->get_position().x,
+			                  p_obs.site->get_position().y);
+			game().send_player_start_or_cancel_expedition(*p_obs.site);
+			start_expedition = false;
 		}
+		p_obs.guard_ship_needed = warships_count / num_ports > p_obs.ships_assigned;
```

Comment:
I think this will distribute warships more evenly: (yes, C integer math is tricky :)

```suggestion
		p_obs.guard_ship_needed = warships_count > num_ports * p_obs.ships_assigned;
```

Though I think there's also more accounting needed for warships, like when ports are lost then any assigned warships should be sent elsewhere, and when a warship is lost, its port's `ships_assigned` should be decreased.

Reply:
well I beleive this would lead to a lot of warships shipping around as it would trigger the guard_ship needed much more often. I assume that integer division is always floored, while in your suggestion it would be ceiled. 
if a port is lost we reduce num_ports and erase the port site observer (defaultai.cc line 6872 by this the ships gets send to an existing port.
the ship does not know to which port it has been send, we only count ships sent to a port, however we might need to check that no ship got send away.

Reply:
> well I beleive this would lead to a lot of warships shipping around as it would trigger the guard_ship needed much more often.

Actually I suspect warships that already have a destination should not look for new assignments. I'm planning to do that in the branch I forked from this, but I hit a hard to debug `std::exception`, so I need to solve that before I can continue testing.

> I assume that integer division is always floored, while in your suggestion it would be ceiled.

Let's see with 3 ports. Warships are ordered until we have 6.

1 ship: ships/ports = 1/3 → 0, no port requests it, so it just lingers around where it was refitted (at least that port gets one ;)  
   with my suggestion all ports request a ship, lucky one gets it  
2 ships: 2/3 → 0: same: if you're lucky, you have 2 ports with 1 ship each, but you may get both ships at the same port  
   my: previous port is happy, another one gets the new ship  
3 ships: 3/3 = 1: finally ports request ships, everything's fine  
   my: 3rd ports gets new ship, everything's fine

Now, if we continue from the same state, we get the same pattern, so let's examine the state where at least one port has 2 ships and at least another (e.g. a newly built one) has 0:

4 ships: 4/3 → 1: port with 2 ships is more than happy, port with 1 ship is happy, though it could request the ship, port without ships requests a ship, 2-1-1 guaranteed  
   my: ports with 1 ship and without ship request ship: this is actually worse, may end up 2-2-0  
5 ships: 5/3 → 1: ports with 2 ships are happy, ports with 1 ship are happy, but shouldn't be, only new port requests ship: may end up 2-1-1 with one ship in limbo, so it may be effectively 3-1-1  
   my: 2-2-1 is guaranteed
6 ships: everything's fine again with both calculations

Extending that to more ports: Your calculation only starts to distribute ships when there are more warships than ports, and after that, ports only request 2nd ship when all ships are ready. With my calculation new ports may lose out if unlucky, but there will always be requests for ready ships. The new port thing could be checked by the ship when choosing a port: if it finds one without ships then go there, otherwise just remember the requesting port, and fulfil the request at the end of the loop (if not broken out by a port without ships).

Reply:
the problem is that ports do not know how many if any are actually there actually. And we iterate over all warships all the time. so we might easily send away a ship in your suggestion and do this much seldom in my current state. 
Again I only intended to implement a real dumb workaround to give the players an impression of the feature.   
however as I am still struggling with time you might easily take over here and just commit to this branch as well.

Reply:
Thank you, I'll do that when I sorted out the above mentioned exception problem.

I also implemented garrisons in my local branch, but that involves more places in the code, so I couldn't make suggestions, and also I'm afraid that the source of the exceptions may be in those changes.